src/LanguageServer/Impl/Implementation/LineFormatter.cs

```diff
@@ -0,0 +1,643 @@
+            var tokenizer = new Tokenizer(languageVersion, options: TokenizerOptions.Verbatim | TokenizerOptions.VerbatimCommentsAndLineJoins | TokenizerOptions.GroupingRecovery);
+            tokenizer.Initialize(reader);
+            _tokenizer = new TokenizerWrapper(tokenizer);
+            _lineTokens = new Dictionary<int, List<TokenExt>>();
+        }
+
+        private void AddToken(TokenExt token) {
+            var line = token.Line;
+
+            // Explicit line joins ("\") appear at the end of a line, but
+            // their span ends on another line, so move backward so they can
+            // be inserted in the right place.
+            if (token.Kind == TokenKind.ExplicitLineJoin) {
+                line--;
+            }
+
+            if (!_lineTokens.TryGetValue(line, out List<TokenExt> tokens)) {
+                tokens = new List<TokenExt>();
+                _lineTokens.Add(line, tokens);
+            }
+
+            tokens.Add(token);
+        }
+
+        /// <summary>
+        /// Tokenizes up to and including the specified line. Tokens are
+        /// stored in _lineTokens. If the provided line number is past the
+        /// end of the input text, then the tokenizer will stop.
+        /// Additionally, this function will attempt to read ahead onto the
+        /// next line to the first non-ignored token so that the formatter
+        /// can look ahead.
+        /// </summary>
+        /// <param name="line">One-indexed line number.</param>
+        /// <returns>A non-null list of tokens on that line.</returns>
+        private List<TokenExt> TokenizeLine(int line) {
+            if (line < 1) {
+                throw new ArgumentOutOfRangeException(nameof(line));
+            }
+
+            var extraToken = true;
+
+            var peeked = _tokenizer.Peek();
+            while (peeked != null && (peeked.Line <= line || extraToken)) {
+                var token = _tokenizer.Next();
+                AddToken(token);
+                peeked = _tokenizer.Peek();
+
+                if (token.Line > line && !token.IsIgnored) {
+                    extraToken = false;
+                }
+            }
+
+            if (!_lineTokens.TryGetValue(line, out List<TokenExt> tokens)) {
+                return new List<TokenExt>();
+            }
+
+            return tokens;
+        }
+
+        /// <summary>
+        /// Formats a single line and returns TextEdits to replace the old text.
+        /// </summary>
+        /// <param name="line">One-indexed line number.</param>
+        /// <returns>A list of TextEdits needed to format the line.</returns>
+        public TextEdit[] FormatLine(int line) {
+            if (line < 1) {
+                return NoEdits;
+            }
+
+            var tokens = TokenizeLine(line);
+
+            // Keep ExplictLineJoin because it has text associated with it.
+            tokens = tokens.Where(t => !t.IsIgnored || t.Kind == TokenKind.ExplicitLineJoin).ToList();
```

Comment:
Looks like we build list, then insta-filter it. Perhaps filter out at creation time in order to reduce duplication? Like pass a filtering lambda to `TokenizeLine`?

Reply:
Perhaps, but I'd be really wary of this since those newline tokens can contain the preceeding whitespace strings which are useful when working out how to edit things. Right now, I've been able to eliminate that need by using the token spans, but it may make future improvements more difficult.

Reply:
@AlexanderSher Any preference for this being `Func<TokenExt, bool>` vs `Predicate<TokenExt>`?

Reply:
Haven't made it to that line yet :)

Reply:
A bit of searching says that Func/Action are the preferred (used in LINQ and others), and Predicate was the original that came with List. I'll probably use Func if there's no objection.